OSOE-113: ⭐ Annotate GitHub repo files with warnings in GitHub workflows

sarahelsaig opened the pull request:
OSOE-113

sarahelsaig committed:
Intentional test failures.

sarahelsaig committed:
enable logger

sarahelsaig committed:
Fix yml

sarahelsaig committed:
UI and Unit tests have different loggers.

sarahelsaig committed:
Actually remove it from unit tests.

sarahelsaig committed:
reverse

sarahelsaig committed:
Alas, it doesn't work.

sarahelsaig committed:
Add annotation writer.

sarahelsaig committed:
Prevent NRE.

sarahelsaig committed:
Access ITest from wrapped test output helper.

sarahelsaig committed:
Fix annotation when exception is in method call.

sarahelsaig committed:
Check out submodules using checkout@v2 instead.

sarahelsaig committed:
SuppressInitialDiagnosticInformation

sarahelsaig committed:
hide "ChromeDriver was started successfully."

sarahelsaig committed:
Reduce test verbosity.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-113

# Conflicts:
#	.github/workflows/build_and_test.yml
#	test/Lombiq.UITestingToolbox

sarahelsaig committed:
restore changes from deleted yml file

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-113

# Conflicts:
#	test/Lombiq.UITestingToolbox

sarahelsaig committed:
Rewrite steps using pwsh.

sarahelsaig committed:
Use ${{ matrix.directory }} as script working directory.

sarahelsaig committed:
bug fix?

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-113

# Conflicts:
#	test/Lombiq.UITestingToolbox

sarahelsaig committed:
Looks like you can't backtick newlines with pwsh?

sarahelsaig committed:
Delete duplicate file.

sarahelsaig committed:
We shouldn't need NoWarn any more.

sarahelsaig committed:
syntax fix

sarahelsaig committed:
Merge remote-tracking branch 'origin/task/workflow-timeout' into issue/OSOE-113

# Conflicts:
#	.github/workflows/build-and-test.yml

sarahelsaig committed:
Use Get-ChildItem.

0liver commented on .github/workflows/build-and-test.yml:
```diff
@@ -34,20 +35,22 @@ jobs:
+          -p:TreatWarningsAsErrors=true `
+          -p:RunAnalyzersDuringBuild=true `
+          -nologo `
+          -consoleLoggerParameters:NoSummary `
```
Trailing whitespace is a backtick killer and may be extremely hard to find.

Here's a long post about the why-s and how-s of avoiding using backticks in the first place. I think [splatting](https://get-powershellblog.blogspot.com/2017/07/bye-bye-backtick-natural-line.html#splattingnote) would be a safe and readable option. What do you think @DAud-IcI and @Piedone?

sarahelsaig replied:
Most (if not all) modern IDEs eliminate trailing whitespace by default so I'm not concerned about that. However escaping newlines (or any non-printing character) is distasteful and makes me throw up a little every time I'm forced to do it I'll give this splatting a try.

0liver replied:
I like the terseness of just using some special character at the end of the line to break commands into many lines, but given the potential problems and thinking about that we want all devs on board with these workflow files, I think that splatting would be a safe approach.

It does make the script look more like code than like a script, but since we're used to reading code all day long anyways, I feel that shouldn't be much of a problem.

Piedone replied:
I don't have an opinion on this.

sarahelsaig committed:
Restore grep -v functionality.

sarahelsaig committed:
Eliminate backticks with splatting.

sarahelsaig committed:
Fix typo

sarahelsaig committed:
Filter out garbage notifications.

sarahelsaig committed:
Is regex actually.

sarahelsaig committed:
Also ignore "ChromeDriver was started successfully." messages.

sarahelsaig committed:
Set screen buffer and update documentation.

sarahelsaig committed:
Grouping goes to Console directly.

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-113

# Conflicts:
#	test/Lombiq.UITestingToolbox

sarahelsaig committed:
Flip all the RUNNER_OS conditions

because Windows is the different one, not Linux. For example most of the commands in the "Linux" branch of the condition should work just fine on macOS and BSD too.

sarahelsaig committed:
Only set BufferWidth on Windows.

sarahelsaig committed:
Suppress notification.

sarahelsaig committed:
Windows is whack.

sarahelsaig committed:
Use Out-String with width.

sarahelsaig committed:
Redirect stderr.

sarahelsaig committed:
Why is the test output missing?

sarahelsaig committed:
Remove unnecessary parentheses.

sarahelsaig committed:
Restore output filters.

sarahelsaig committed:
Use temporary file.

sarahelsaig committed:
No tests were run.

sarahelsaig committed:
Remove accidentally left-in echo.

sarahelsaig committed:
Let's export test.out to know for sure if that's where text gets wrapped.

sarahelsaig committed:
Using bash to output.

sarahelsaig committed:
Fix broken pipe.

sarahelsaig committed:
Package artifact after returning to ITestOutputHelper usage.

sarahelsaig committed:
unusing

sarahelsaig committed:
Fix Windows ConnectionStringTemplate.

sarahelsaig committed:
Remove IntentionalTestErrors

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-113

# Conflicts:
#	test/Lombiq.UITestingToolbox

sarahelsaig committed:
Add note in readme.

Piedone committed:
Deliberately failing test to see workflow annotations on GitHub

Piedone commented on .github/workflows/build-and-test.yml:
```diff
@@ -35,52 +35,59 @@ jobs:
         {
-            docker pull mcr.microsoft.com/mssql/server &&
-            docker run --name sql2019 -e 'ACCEPT_EULA=Y' -e 'SA_PASSWORD=Password1!' -p 1433:1433 -d 'mcr.microsoft.com/mssql/server:2019-latest' &&
-            docker exec -u 0 sql2019 bash -c 'mkdir /data; chmod 777 /data -R; chown mssql:root /data'
+            choco install sql-server-express --no-progress
         }
         else
         {
-            choco install sql-server-express --no-progress
+            docker pull mcr.microsoft.com/mssql/server &&
+            docker run --name sql2019 -e 'ACCEPT_EULA=Y' -e 'SA_PASSWORD=Password1!' -p 1433:1433 -d 'mcr.microsoft.com/mssql/server:2019-latest' &&
+            docker exec -u 0 sql2019 bash -c 'mkdir /data; chmod 777 /data -R; chown mssql:root /data'
         }
```
Why the swap here and below?

sarahelsaig replied:
Because Windows is the weird one. If we need to add any other OS likely the Linux commands should work.

Piedone commented on .github/workflows/build-and-test.yml:
```diff
@@ -104,40 +111,66 @@ jobs:
-    # produced by chromedriver. The trouble is, this way the final error code will be grep's and not the test run's, so
-    # to get around this 'SUCCESS' or 'FAILURE' is written into the .ui_test_result file before the output is sent to
-    # grep. The final line sets the result according to what was written into the .ui_test_result file.
+    # First we globally set test configurations using environment variables. Then acquire the list of all test projects
+    # (excluding the two test libraries) and then runs each until one fails or all concludes. If a test fails the output
+    # is sanitized from unnecessary diagnostics messages from chromedriver if the output doesn't already contain
+    # groupings then it wraps them in '::group::<project name>'. If there are already groupings, then this is not
+    # possible to nest them (https://github.com/actions/runner/issues/802) so that's omitted. The groupings make the
+    # output collapsible region on the Actions web UI. Note that we use bash to output the log using bash to avoid pwsh
+    # wrapping the output to the default buffer width.
```
May need some line length formatting:

```suggestion
    # First, we globally set test configurations using environment variables. Then acquire the list of all test projects
    # (excluding the two test libraries) and then run each until one fails or all concludes. If a test fails, the output
    # is sanitized from unnecessary diagnostics messages from chromedriver if the output doesn't already contain
    # groupings, then it wraps them in '::group::<project name>'. If there are already groupings, then it is not
    # possible to nest them (https://github.com/actions/runner/issues/802) so that's omitted. The groupings make the
    # output collapsible region on the Actions web UI. Note that we use bash to output the log using bash to avoid pwsh
    # wrapping the output to the default buffer width.
```

Piedone committed:
Revert "Deliberately failing test to see workflow annotations on GitHub"

This reverts commit 081d8ca1864132f3a8563fa60a208ec6d78e95fe.

sarahelsaig committed:
Update .github/workflows/build-and-test.yml

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

sarahelsaig committed:
Edge SuppressInitialDiagnosticInformation

sarahelsaig committed:
update submodules

sarahelsaig committed:
Merge remote-tracking branch 'origin/dev' into issue/OSOE-113

# Conflicts:
#	.github/workflows/build-and-test.yml
#	.gitignore

Piedone committed:
Updating UITestingToolbox to merged dev
